test/rules/avoid_void_async.dart

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N avoid_void_async`
+import 'dart:async';
+
+void main() {
+  () {}; // OK
+  () => null; // OK
+  () async {}; // OK
+  () async => null; // OK
+}
+
+void a() async {} // LINT
+void b() {} // OK
+Future<void> c() async {} // OK
+void d() async => null; // LINT
+void e() => null; // OK
+Future<void> f() async => null; // OK
+
+typedef void f1(int x); // OK
+
+typedef Future<void> f2(int x); // OK
+
+class Foo {
+  void a() async {} // LINT
+  void b() {} // OK
+  Future<void> c() async {} // OK
+  void d() async => null; // LINT
+  void e() => null; // OK
+  Future<void> f() async => null; // OK
```

Comment:
This is missing some test cases that I think it would be good to add. I can think of the following:

- declarations marked with `async*`,
- class members marked with `static`,
- getters, setters, and operators.

Also, it might be good to add one declaration marked with `sync` and one with `sync*` (I wouldn't necessarily duplicate those examples everywhere.

Reply:
It looks like setters cannot be asynchronous...the previous setter code was there from always_annotate_return_types.

Added tests for async*, sync*, non async setters, and sync/async void getters & operators, and static.

Great suggestions, thank you Brian! :)